### PR TITLE
aws - asg refactor for launch-template support

### LINKED ISSUE #3252: aws.asg - launch template filter and support in not-encypted filter for launch templates
autoscaling groups support using ec2 launch templates in lieu of launch configurations. Launch templates should be a new resource type. For asgs, the launch-template should be an additional filter, and the unencrypted filter should support both source of instance configuration when evaluating.

### LINKED ISSUE #3483: ami unused filter fails with auto scaling group using launch template vs launch config
I'm seeing this traceback for an ami resource policy using the unused:true filter.
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/c7n/policy.py", line 232, in run
    resources = self.policy.resource_manager.resources()
  File "/usr/local/lib/python3.5/dist-packages/c7n/resources/ami.py", line 60, in resources
    return super(AMI, self).resources(query=query)
  File "/usr/local/lib/python3.5/dist-packages/c7n/query.py", line 413, in resources
    resources = self.filter_resources(resources)
  File "/usr/local/lib/python3.5/dist-packages/c7n/manager.py", line 105, in filter_resources
    resources = f.process(resources, event)
  File "/usr/local/lib/python3.5/dist-packages/c7n/filters/core.py", line 270, in process
    resources = f.process(resources, events)
  File "/usr/local/lib/python3.5/dist-packages/c7n/resources/ami.py", line 250, in process
    images = self._pull_ec2_images().union(self._pull_asg_images())
  File "/usr/local/lib/python3.5/dist-packages/c7n/resources/ami.py", line 239, in _pull_asg_images
    lcfgs = set(a['LaunchConfigurationName'] for a in asgs)
  File "/usr/local/lib/python3.5/dist-packages/c7n/resources/ami.py", line 239, in <genexpr>
    lcfgs = set(a['LaunchConfigurationName'] for a in asgs)
KeyError: 'LaunchConfigurationName'
```
The query of autoscaling groups is returning resources using launch templates, not launch configs.  The launch template resource has no 'LaunchConfigurationName' key.

```
{
    "AutoScalingGroups": [
        {
            "HealthCheckGracePeriod": 300,
            "DesiredCapacity": 1,
            "DefaultCooldown": 300,
            "AvailabilityZones": [
                "us-east-2a"
            ],
            "ServiceLinkedRoleARN": "arn:aws:iam::<account>:role/aws-service-role/autoscaling.amazonaws.com/AWSServiceRoleForAutoScaling",
            "CreatedTime": "2018-10-07T12:27:48.977Z",
            "EnabledMetrics": [],
            "AutoScalingGroupName": "test-app-asg",
            "TerminationPolicies": [
                "Default"
            ],
            "AutoScalingGroupARN": "arn:aws:autoscaling:us-east-2:<account>:autoScalingGroup:50f5-4673-ac27-94e6be7a61:autoScalingGroupName/test-app-asg",
            "HealthCheckType": "ELB",
            "Instances": [
                {
                    "ProtectedFromScaleIn": false,
                    "AvailabilityZone": "us-east-2a",
                    "HealthStatus": "Healthy",
                    "LifecycleState": "InService",
                    "InstanceId": "i-029395942a39f",
                    "LaunchTemplate": {
                        "LaunchTemplateName": "test-app-launch-template",
                        "LaunchTemplateId": "lt-03358423448c4",
                        "Version": "5"
                    }
                }
            ],
            "MaxSize": 1,
            "NewInstancesProtectedFromScaleIn": false,
            "LoadBalancerNames": [],
            "VPCZoneIdentifier": "subnet-0e8b85423443",
            "SuspendedProcesses": [],
            "Tags": [
            ],
            "TargetGroupARNs": [
                "arn:aws:elasticloadbalancing:us-east-2:<account>:targetgroup/test-app-lb-target-group/754eeffc"
            ],
            "MinSize": 1,
            "LaunchTemplate": {
                "LaunchTemplateName": "test-app-launch-template",
                "LaunchTemplateId": "lt-03358423",
                "Version": "$Latest"
            }
        }
]

```
custodian version: 0.8.32.1

### PR DESCRIPTION
**Pull request by kapilt**
[updated] This introduces comprehensive support in asg resource and filters for launch template, also unifies asg filter handling of ebs snapshots and amis. The later enables some nice performance wins for accounts with large snapshot volume, while the ami improvements, take advantage of some newer api behavior. 

This closes #3483  and closes #3252

### CODE REVIEW

**Comment by kapilt** [comment]
thinking it may be closer to desired to model launch templates as launch template versions.

**Comment by kapilt** [comment]
remaining bits are to switch all the use of launch config as a mixing, to using composition, and then some additional tests with templates.

**Comment by kapilt** [comment]
noting this is still work in progress, i'm not particularly happy with the current approach, it still has too much complexity and overlap between launch configs and launch templates. Going to look at a more fundamental refactor of asg filters.

**Comment by kapilt** [comment]
current state for this looks pretty good, going to do a final review on it tomorrow, but the refactoring to a launch info to account for delta between launch templates and launch configurations worked out nicely.

File: .gitignore [hunk 1]
```diff
@@ -32,5 +32,7 @@ results.xml
 .vscode/
 !.vscode/*.sample
 .pytest_cache/
+<<<<<<< HEAD
```
Comment (kapilt) [inline .gitignore]: stray conflict marker

**Comment by dschro-1993** [comment]
Hey @kapilt :)
How do I check for launch templates with public network settings on ASGs?
I think we need to enhance the docs here how to access LT-data.

```
    filters:
      - or:
        - type:  launch-config
          key:   AssociatePublicIpAddress
          value: true
        - type:  launch-config # Launch-Template
          key:   "[].NetworkInterfaces[].AssociatePublicIpAddress"
          value: true
```

This does not work.

**Comment by dschro-1993** [comment]
It seems like the `launch-config` filter does not support JMESPath expressions, because this one works:

```
filters:
  - type:  launch-config
    key:   "InstanceType"
    value: "t2.micro"
```

**Pull request merged**